File: src/liveplugin/pluginrunner/PluginRunner.java
package liveplugin.pluginrunner;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.util.Function;
import com.intellij.util.containers.ContainerUtil;
import groovy.lang.GroovyClassLoader;
import org.apache.oro.io.GlobFilenameFilter;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public interface PluginRunner {
	String IDE_STARTUP = "IDE_STARTUP";
	String ADD_TO_CLASSPATH_KEYWORD = "add-to-classpath ";
	String DEPENDS_ON_PLUGIN_KEYWORD = "depends-on-plugin ";

	/**
	 * @param pathToPluginFolder absolute path to plugin folder
	 * @return true if {@link PluginRunner} can run plugin in this folder
	 */
	boolean canRunPlugin(String pathToPluginFolder);

	/**
	 * @param pathToPluginFolder absolute path to plugin folder
	 * @param pluginId plugin id, e.g. to distinguish it from other plugins in error messages
	 * @param binding map with implicit variables available in plugin script
	 * @param runOnEDTCallback callback which should be used to run plugin code on EDT
	 */
	void runPlugin(String pathToPluginFolder, String pluginId, Map<String, ?> binding, Function<Runnable, Void> runOnEDTCallback);

	String scriptName();

	class ClasspathAddition {
		private static final Logger LOG = Logger.getInstance(ClasspathAddition.class);

		public static ClassLoader createClassLoaderWithDependencies(List<String> pathsToAdd, String mainScriptUrl, String pluginId, ErrorReporter errorReporter) {
			GroovyClassLoader classLoader = new GroovyClassLoader(PluginRunner.class.getClassLoader());
			try {

				for (String path : pathsToAdd) {
					if (path.startsWith("file:/")) {
						URL url = new URL(path);
						classLoader.addURL(url);
						classLoader.addClasspath(url.getFile());
					} else {
						classLoader.addURL(new URL("file:///" + path));
						classLoader.addClasspath(path);
					}
				}

			} catch (IOException e) {
				errorReporter.addLoadingError(pluginId, "Error while looking for dependencies in '" + mainScriptUrl + "'. " + e.getMessage());
			}
			return classLoader;
		}

		public static List<String> findClasspathAdditions(String[] lines, String prefix, Map<String, String> environment, Function<String, Void> onError) throws IOException {
			List<String> pathsToAdd = new ArrayList<String>();
			for (String line : lines) {
				if (!line.startsWith(prefix)) continue;

				String path = line.replace(prefix, "").trim();
				path = inlineEnvironmentVariables(path, environment);

				List<String> matchingFiles = findMatchingFiles(path);
				if (matchingFiles.isEmpty()) {
					onError.fun(path);
				} else {
					pathsToAdd.addAll(matchingFiles);
				}
			}
			return pathsToAdd;
		}

		private static List<String> findMatchingFiles(String pathAndPattern) {
			if (new File(pathAndPattern).exists()) return Collections.singletonList(pathAndPattern);

			int separatorIndex = pathAndPattern.lastIndexOf(File.separator);
			String path = pathAndPattern.substring(0, separatorIndex + 1);
			String pattern = pathAndPattern.substring(separatorIndex + 1);

			File[] files = new File(path).listFiles((FileFilter) new GlobFilenameFilter(pattern));
			files = (files == null ? new File[0] : files);
			return ContainerUtil.map(files, new Function<File, String>() {
				@Override public String fun(File file) {
					return file.getAbsolutePath();
				}
			});
		}

		private static String inlineEnvironmentVariables(String path, Map<String, String> environment) {
			boolean wasModified = false;
			for (Map.Entry<String, String> entry : environment.entrySet()) {
				path = path.replace("$" + entry.getKey(), entry.getValue());
				wasModified = true;
			}
			if (wasModified) LOG.info("Additional classpath with inlined env variables: " + path);
			return path;
		}
	}
}

File: src/main/liveplugin/pluginrunner/PluginRunner.kt
package liveplugin.pluginrunner

import com.intellij.ide.plugins.IdeaPluginDescriptor
import com.intellij.ide.plugins.IdeaPluginDescriptorImpl
import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.extensions.DefaultPluginDescriptor
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.util.lang.UrlClassLoader
import groovy.lang.GroovyClassLoader
import liveplugin.Result
import liveplugin.asFailure
import liveplugin.asSuccess
import liveplugin.pluginrunner.AnError.LoadingError
import liveplugin.toUrl
import org.apache.oro.io.GlobFilenameFilter
import java.io.File
import java.io.FileFilter
import java.util.*

interface ExecutablePlugin

interface PluginRunner {
    val scriptName: String

    fun setup(plugin: LivePlugin, project: Project?): Result<ExecutablePlugin, AnError>

    fun run(executablePlugin: ExecutablePlugin, binding: Binding): Result<Unit, AnError>

    object ClasspathAddition {
        fun createClassLoaderWithDependencies(
            additionalClasspath: List<File>,
            pluginDescriptors: List<IdeaPluginDescriptor>,
            plugin: LivePlugin
        ): Result<ClassLoader, LoadingError> {
            val classLoader = GroovyClassLoader(createParentClassLoader(pluginDescriptors, plugin))

            additionalClasspath.forEach { file ->
                if (!file.exists()) return LoadingError("Didn't find plugin dependency '${file.absolutePath}'.").asFailure()
            }
            additionalClasspath.forEach { file -> classLoader.addURL(file.toUrl()) }

            return classLoader.asSuccess()
        }

        private fun createParentClassLoader(pluginDescriptors: List<IdeaPluginDescriptor>, plugin: LivePlugin): ClassLoader {
            val parentLoaders = pluginDescriptors.mapNotNull { it.pluginClassLoader } + PluginRunner::class.java.classLoader
            return PluginClassLoader_Fork(
                UrlClassLoader.build().files(emptyList()).allowLock(true).useCache(),
                parentLoaders.toTypedArray(),
                DefaultPluginDescriptor(plugin.id),
                null,
                PluginManagerCore::class.java.classLoader,
                null,
                null
            )
        }

        fun findPluginDescriptorsOfDependencies(lines: List<String>, keyword: String): List<Result<IdeaPluginDescriptor, String>> {
            return lines.filter { line -> line.startsWith(keyword) }
                .map { line -> line.replace(keyword, "").trim { it <= ' ' } }
                .map { PluginManagerCore.getPlugin(PluginId.getId(it))?.asSuccess() ?: "Failed to find dependent plugin '$it'.".asFailure() }
        }

        fun List<IdeaPluginDescriptor>.withTransitiveDependencies(): List<IdeaPluginDescriptor> {
            val result = HashSet<IdeaPluginDescriptor>()
            val queue = LinkedList(this)
            while (queue.isNotEmpty()) {
                val descriptor = queue.remove()
                if (descriptor !in result) {
                    result.add(descriptor)

                    val dependenciesDescriptors = descriptor.dependencies.mapNotNull {
                        if (it.isOptional) null else PluginManagerCore.getPlugin(it.pluginId)
                    }

                    @Suppress("UnstableApiUsage") // This is a "temporary" hack for https://youtrack.jetbrains.com/issue/IDEA-206274
                    val dependenciesDescriptors2 =
                        if (descriptor !is IdeaPluginDescriptorImpl) emptyList()
                        else descriptor.dependencies.plugins.mapNotNull { PluginManagerCore.getPlugin(it.id) }

                    queue.addAll((dependenciesDescriptors + dependenciesDescriptors2).distinctBy { it.pluginId })
                }
            }
            return result.toList()
        }

        fun findClasspathAdditions(lines: List<String>, keyword: String, environment: Map<String, String>): List<Result<List<File>, String>> {
            return lines.filter { line -> line.startsWith(keyword) }
                .map { line -> line.replace(keyword, "").trim { it <= ' ' } }
                .map { line -> line.inlineEnvironmentVariables(environment) }
                .map { path ->
                    val files = findMatchingFiles(path).map { File(it) }
                    if (files.isEmpty()) path.asFailure() else files.asSuccess()
                }
        }

        private fun findMatchingFiles(pathAndPattern: String): List<String> {
            if (File(pathAndPattern).exists()) return listOf(pathAndPattern)

            val separatorIndex = pathAndPattern.lastIndexOf(File.separator)
            val path = pathAndPattern.substring(0, separatorIndex + 1)
            val pattern = pathAndPattern.substring(separatorIndex + 1)

            val files = File(path).listFiles(GlobFilenameFilter(pattern) as FileFilter) ?: emptyArray()
            return files.map { it.absolutePath }
        }

        private fun String.inlineEnvironmentVariables(environment: Map<String, String>): String {
            var result = this
            environment.forEach { (key, value) ->
                result = result.replace("$$key", value)
            }
            return result
        }
    }
}

sealed class AnError {
    data class LoadingError(val message: String = "", val throwable: Throwable? = null) : AnError()
    data class RunningError(val throwable: Throwable) : AnError()
}
